catalogd/internal/storage/localdir.go

```diff
@@ -2,113 +2,316 @@ package storage
+	return filepath.Join(catalogDir, "index.json")
+}
+
+type storeMetasFunc func(catalogDir string, metaChan <-chan *declcfg.Meta) error
+
+func storeCatalogData(catalogDir string, metas <-chan *declcfg.Meta) error {
+	f, err := os.Create(catalogFilePath(catalogDir))
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	for m := range metas {
+		if _, err := f.Write(m.Blob); err != nil {
+			return err
+		}
+	}
+	return nil
 }
 
-func (s LocalDirV1) Delete(catalog string) error {
-	return os.RemoveAll(filepath.Join(s.RootDir, catalog))
+func storeIndexData(catalogDir string, metas <-chan *declcfg.Meta) error {
+	idx := newIndex(metas)
+
+	f, err := os.Create(catalogIndexFilePath(catalogDir))
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	enc := json.NewEncoder(f)
+	enc.SetEscapeHTML(false)
+	return enc.Encode(idx)
 }
 
-func (s LocalDirV1) BaseURL(catalog string) string {
+func (s *LocalDirV1) BaseURL(catalog string) string {
 	return s.RootURL.JoinPath(catalog).String()
 }
 
-func (s LocalDirV1) StorageServerHandler() http.Handler {
+func (s *LocalDirV1) StorageServerHandler() http.Handler {
 	mux := http.NewServeMux()
-	fsHandler := http.FileServer(http.FS(&filesOnlyFilesystem{os.DirFS(s.RootDir)}))
-	spHandler := http.StripPrefix(s.RootURL.Path, fsHandler)
-	gzHandler := gzhttp.GzipHandler(spHandler)
 
-	typeHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+	mux.HandleFunc(s.RootURL.JoinPath("{catalog}", "api", "v1", "all").Path, s.handleV1All)
+	if s.EnableQueryHandler {
+		mux.HandleFunc(s.RootURL.JoinPath("{catalog}", "api", "v1", "query").Path, s.handleV1Query)
+	}
+	allowedMethodsHandler := func(next http.Handler, allowedMethods ...string) http.Handler {
+		allowedMethodSet := sets.New[string](allowedMethods...)
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			if !allowedMethodSet.Has(r.Method) {
+				http.Error(w, http.StatusText(http.StatusMethodNotAllowed), http.StatusMethodNotAllowed)
+				return
+			}
+			next.ServeHTTP(w, r)
+		})
+	}
+	return allowedMethodsHandler(mux, http.MethodGet, http.MethodHead)
+}
+
+func (s *LocalDirV1) handleV1All(w http.ResponseWriter, r *http.Request) {
+	s.m.RLock()
+	defer s.m.RUnlock()
+
+	catalog := r.PathValue("catalog")
+	catalogFile, catalogStat, err := s.catalogData(catalog)
+	if err != nil {
+		httpError(w, err)
+		return
+	}
+	w.Header().Add("Content-Type", "application/jsonl")
+	http.ServeContent(w, r, "", catalogStat.ModTime(), catalogFile)
+}
+
+func (s *LocalDirV1) handleV1Query(w http.ResponseWriter, r *http.Request) {
+	s.m.RLock()
+	defer s.m.RUnlock()
+
+	catalog := r.PathValue("catalog")
+	catalogFile, catalogStat, err := s.catalogData(catalog)
+	if err != nil {
+		httpError(w, err)
+		return
+	}
+	defer catalogFile.Close()
+
+	w.Header().Set("Last-Modified", catalogStat.ModTime().UTC().Format(TimeFormat))
+	if checkPreconditions(w, r, catalogStat.ModTime()) {
+		w.WriteHeader(http.StatusNotModified)
```

Comment:
It looks like `checkPreconditions` already sets the appropriate status code, so no need to do that here?

(This may be a bug because `checkPreconditions` can also write the `http.StatusPreconditionFailed` status code)